src/twisted/web/proxy.py

```diff
@@ -288,7 +288,7 @@ def render(self, request):
         if self.port == 80:
             host = self.host
         else:
-            host = self.host + u":" + str(self.port)
+            host = u"%s:%d" % (self.host, self.port)
```

Comment:
If a hostname is given to us as a `str`, it should probably be encoded using `idna` rather than `ascii`. (Unless the HTTP RFC has some other requirement).

Maybe something along the lines of 

```
host = self.host.encode('idna')
if self.port != 80:
    host += b':' + str(self.port).encode('ascii')
```

Reply:
> If a hostname is given to us as a `str`, it should probably be encoded using `idna` rather than `ascii`. (Unless the HTTP RFC has some other requirement).
I wouldn't dare to mess with encodings. And AFAIU hostname is required to be unicode and the whole code would crash elsewhere otherwise. In any case I belive this should be handled at the webserver constructor or alike, not in the resource `render` method.

> or `host += b':%d' % (self.port,)` if you prefer format strings.
That would be elegant but beware that `host != self.host`

I don't really have a strong interest in changing these lines. I just felt the `+ u":" +` concatenation was somewhat creepy and a string format would look prettier. But this is a completely optional and opinionated change!

Reply:
> Yet I think this should be done elsewhere and is beyond the purpose of this PR

Looking at this further, the modern IDNA encoder is only a soft-depndency of Twisted, and this isn't the only place in this module where unicode hostnames are converted to a byte representation, so I agree this is a complicated enough change that it should be in its own PR. Sorry for the distraction!

> Can you point me to any non-ascii standard website to test it?

For curiosity's sake, here are a few: `http://santé.gouv.qc.ca/` (encodes as `xn--sant-epa.gouv.qc.ca`), `www.köln.de` (`www.xn--kln-sna.de`, but the site redirects to `www.koeln.de`).